### PR TITLE
CreateTemplatePartModalContents: use native radio inputs

### LINKED ISSUE #66569: CreateTemplatePartModalContents: Refactor to not use `RadioGroup`
## What problem does this address?

The `RadioGroup` component is deprecated (#45389), and this is the only instance of usage in the Gutenberg app. It's also heavily modified visually. It doesn't make sense to keep this around. Refactoring this usage away from RadioGroup will allow us to potentially soft or hard deprecate (#61099) RadioGroup.

Heavily modded RadioGroup

## What is your proposed solution?

Use standard HTML instead of the RadioGroup component.

### PR DESCRIPTION
**Pull request by ciampo**
## What?

Fixes #66569

Refactor the `CreateTemplatePartModalContents` component to use native `<input type="radio" />` elements instead of the deprecated `RadioGroup` component

## Why?

The `RadioGroup` component is deprecated and we shouldn't use it in the repository.

This refactor brings in other several benefits:

- using native HTML elements ensures more semantic markup:
  - radios are grouped via `fieldset` and `legend`;
  - radios are properly labelled and described;
  - the radio group is a composite widget, and implements the correct keyboard navigation;
- small styles tweaks:
  - the radio group borders now match the ones of the text input in the same modal;
  - cleaner grid layout, better spacing;
- less style overrides and internal component classnames

## How?

- Removed the old code, and replaced it with semantic markup (`fieldset`, `legend`, `radio` and `label`)
- Tweaked the styles to meet the target design

## Testing Instructions

- In the site editor, visit the "Patterns" page
- Click the "Add new pattern" button in the top right part of the screen, and select "Add new template pattern" — a modal dialog should show;
- Make sure that the "Area" radio group in the modal (and the modal in general) looks and works like on `trunk` (apart from the differences listed above)

## Screenshots or screencast 

|Before|After|
|-|-|
| Screenshot 2024-12-07 at 00 57 00 | Screenshot 2024-12-07 at 01 01 06 |
| <video src="https://github.com/user-attachments/assets/e9cf3251-c22d-440b-a065-a6c9cccdd7d3" /> | <video src="https://github.com/user-attachments/assets/330a137f-c989-4dfb-8bf0-d4dc44af2bb1" /> |

### CODE REVIEW

File: packages/fields/src/components/create-template-part-modal/index.tsx [hunk 1]
```diff
@@ -77,7 +71,6 @@ export default function CreateTemplatePartModal( {
 			focusOnMount="firstContentElement"
 			size="medium"
 		>
-			{ /* @ts-ignore */ }
```
Comment (ciampo) [inline packages/fields/src/components/create-template-part-modal/index.tsx]: I'll remove this line by rebasing on `trunk` once #67709 is merged

**Comment by jameskoster** [comment]
Good change. 

Probably not something to discuss here, but we could potentially remove all the customisations and use the vanilla component...

Screenshot 2024-12-09 at 12 37 03

Either that, or formalise this design treatment as a variant so it can be reused elsewhere.

File: packages/fields/src/components/create-template-part-modal/index.tsx [hunk 2]
```diff
@@ -40,6 +34,13 @@ import {
 	useExistingTemplateParts,
 } from './utils';
 
+function getAreaRadioId( value: string ) {
+	return `fields-create-template-part-modal__area-option-${ value }`;
+}
+function getAreaRadioDescriptionId( value: string ) {
+	return `fields-create-template-part-modal__area-option-description-${ value }`;
+}
```
Comment (mirka) [inline packages/fields/src/components/create-template-part-modal/index.tsx]: Seeing how `CreateTemplatePartModalContents` is exported to multiple places, I'm not 100% sure that this fieldset will never been rendered twice in the same document 😅 Is it worth adding in an instance id?

File: packages/fields/src/components/create-template-part-modal/index.tsx [hunk 3]
```diff
@@ -203,51 +202,60 @@ export function CreateTemplatePartModalContents( {
...
-						onChange={ ( value ) =>
-							value && typeof value === 'string'
-								? setArea( value )
-								: () => void 0
-						}
-						checked={ area }
+				<fieldset>
+					<BaseControl.VisualLabel
+						as="legend"
+						className="fields-create-template-part-modal__area-radio-group-label"
```
Comment (mirka) [inline packages/fields/src/components/create-template-part-modal/index.tsx]: Unused class?

File: packages/fields/src/components/create-template-part-modal/index.tsx [hunk 4]
```diff
@@ -203,51 +202,60 @@ export function CreateTemplatePartModalContents( {
...
-											{ area === value && (
-												<Icon icon={ check } />
-											) }
-										</FlexItem>
-									</Flex>
-								</Radio>
+									<input
+										type="radio"
+										id={ getAreaRadioId( value ) }
+										name="fields-create-template-part-modal__area"
```
Comment (mirka) [inline packages/fields/src/components/create-template-part-modal/index.tsx]: I think you meant `className` here, but also it looks unused?

File: packages/fields/src/components/create-template-part-modal/style.scss [hunk 5]
```diff
@@ -3,61 +3,86 @@
...
+	input[type="radio"]:not(:checked) ~ & {
+		opacity: 0;
+	}
+}
+
+.fields-create-template-part-modal__area-radio-description {
+	grid-column: 2 / 3;
+	margin: 0;
+
+	color: $gray-600;
```
Comment (mirka) [inline packages/fields/src/components/create-template-part-modal/style.scss]: Looks like this is a preexisting issue, but I feel this is too light. The descriptions in BaseControl (and thus RadioControl) are gray-700, so I think we should go with that?

File: packages/fields/src/components/create-template-part-modal/style.scss [hunk 6]
```diff
@@ -3,61 +3,86 @@
...
+
+.fields-create-template-part-modal__area-radio-icon,
+.fields-create-template-part-modal__area-radio-checkmark {
+	fill: currentColor;
+}
+
+.fields-create-template-part-modal__area-radio-checkmark {
+	input[type="radio"]:not(:checked) ~ & {
+		opacity: 0;
+	}
```
Comment (mirka) [inline packages/fields/src/components/create-template-part-modal/style.scss:73]: 😎

**Review by mirka — COMMENTED** [review]
Thanks for the cleanup!

**Comment by mirka** [comment]
By the way I'm getting a "The 'useSelect' hook returns different values when called with the same state and parameters" warning when opening the modal, but confirmed it's already an issue in trunk.

**Comment by jsnajdr** [comment]
> By the way I'm getting a "The 'useSelect' hook returns different values when called with the same state and parameters" warning when opening the modal

This caught my attention 🙂 and I believe I have a fix in #67786. Please try to apply the fix and let me know if it eliminates the warning for you. I didn't do much testing myself 🙂

**Comment by mirka** [comment]
> we could potentially remove all the customisations and use the vanilla component

I love this idea. #66615 will be a prerequisite for this.

File: packages/fields/src/components/create-template-part-modal/index.tsx (hunk 4 above)
Comment (ciampo) [reply packages/fields/src/components/create-template-part-modal/index.tsx]: Shouldn't all radios in the same radio group share the same `name` ?

**Comment by ciampo** [comment]
> Probably not something to discuss here, but we could potentially remove all the customisations and use the vanilla component [...] Either that, or formalise this design treatment as a variant so it can be reused elsewhere.

My preference would be to remove the customization, instead of introducing a new variant.

Does this mean that we should stop working on this PR, wait for #66615 and then refactor using `RadioControl` instead?

**Comment by tyxla** [comment]
> > By the way I'm getting a "The 'useSelect' hook returns different values when called with the same state and parameters" warning when opening the modal
> 
> This caught my attention 🙂 and I believe I have a fix in #67786. Please try to apply the fix and let me know if it eliminates the warning for you. I didn't do much testing myself 🙂

#67786 landed already, so the warning should disappear after rebasing to the latest `trunk`.

**Comment by jameskoster** [comment]
I'd probably lean towards proceeding with the PR as-is, to avoid changing the shipping UI. 

We can then revisit the design in a dedicated issue and evaluate:

1. Whether using the vanilla component styles would be detrimental to the design (I don't really think so, but that needs more feedback from other designers).
2. Whether a different control (`ToggleGroupControl`) might be a better fit.

File: packages/fields/src/components/create-template-part-modal/index.tsx (hunk 4 above)
Comment (mirka) [reply packages/fields/src/components/create-template-part-modal/index.tsx]: Lol sorry, you're right. Same comment about maybe using an instance id applies then.

File: packages/fields/src/components/create-template-part-modal/index.tsx (hunk 3 above)
Comment (ciampo) [reply packages/fields/src/components/create-template-part-modal/index.tsx]: https://github.com/WordPress/gutenberg/pull/67702/commits/ebce71f647d81385fdc74bbbae6dee89c6e394bd

File: packages/fields/src/components/create-template-part-modal/index.tsx (hunk 4 above)
Comment (ciampo) [reply packages/fields/src/components/create-template-part-modal/index.tsx]: https://github.com/WordPress/gutenberg/pull/67702/commits/16fd5321f39899713d7546d869d505ea9d23d6be

File: packages/fields/src/components/create-template-part-modal/index.tsx (hunk 2 above)
Comment (ciampo) [reply packages/fields/src/components/create-template-part-modal/index.tsx]: https://github.com/WordPress/gutenberg/pull/67702/commits/16fd5321f39899713d7546d869d505ea9d23d6be

File: packages/fields/src/components/create-template-part-modal/style.scss (hunk 5 above)
Comment (ciampo) [reply packages/fields/src/components/create-template-part-modal/style.scss]: https://github.com/WordPress/gutenberg/pull/67702/commits/34a93cdbe4078874f0470c7d9e0eb21fcf987fa4

**Comment by ciampo** [comment]
Rebased and addressed all feedback.

I'm good with merging this PR and then iterating with a new design in a separate issue.

**Review by jameskoster — APPROVED** [review]

**Pull request merged**